### PR TITLE
feat: no longer use tsconfck

### LINKED ISSUE #878: Production Dependency Includes Typescript
**Describe the bug**
It seems that the last few versions of `@oclif/core` has re-introduced TypeScript as a production dependency, significantly increase the bundle size of CLIs using OCLIF.  This appears to be due to the `tsconfck` dependency

```
└─┬ @oclif/core@3.10.8
  └─┬ tsconfck@3.0.0
    └── typescript@5.2.2
```

This seems to increase bundle size by about 10MB depending on compression technique and OS target.

**To Reproduce**
1. Build an OCLIF CLI using the latest version
2. Inspect the bundles

**Expected behavior**
TypeScript is not a production dependency.

**Screenshots**
N/A

**Environment (please complete the following information):**
 - OS & version: N/A
 - Shell/terminal & version: N/A

**Additional context**
N/A

### LINKED ISSUE #907: tsconfck module cannot be found
**Describe the bug**
`tsconfck` as a module cannot be found for some reason.

Originally discovered when this workflow started failing in one of our repos: https://github.com/mitre/saf_action/actions/runs/7285825454/job/19853410684.  The specific issue there was that jest started segfaulting.  Tracing the code showed that the crash occurred when trying to dynamically import `tsconfck` [here](https://github.com/oclif/core/blob/main/src/util/fs.ts#L76).

I originally thought that the issue might have had something to do with using dynamic imports, so I tried to statically import the library using the following patch:

<details>

<summary>
patch to src/util/fs.ts
</summary>

```
diff --git a/src/util/fs.ts b/src/util/fs.ts
index 0e09e7c..3b5022f 100644
--- a/src/util/fs.ts
+++ b/src/util/fs.ts
@@ -1,6 +1,7 @@
 import {Stats, existsSync as fsExistsSync, readFileSync} from 'node:fs'
 import {readFile, stat} from 'node:fs/promises'
 import {join} from 'node:path'
+import {parse as tsparse} from 'tsconfck'

 import {mergeNestedObjects} from './util'

@@ -73,8 +74,7 @@ export function existsSync(path: string): boolean {
 }

 export async function readTSConfig(path: string) {
-  const {parse} = await import('tsconfck')
-  const result = await parse(path)
+  const result = await tsparse(path)
   const tsNodeOpts = mergeNestedObjects(result.extended ?? [result], 'tsconfig.ts-node')
   return {...result.tsconfig, 'ts-node': tsNodeOpts}
 }
```

</details>

When trying to build the project, it gave the following error: `Cannot find module 'tsconfck' or its corresponding type declarations.`.

<details>
<summary>
Full `yarn pack` output
</summary>

```
$ yarn pack
yarn pack v1.22.21
$ yarn run build
yarn run v1.22.21
$ shx rm -rf lib && tsc
src/util/fs.ts:4:32 - error TS2307: Cannot find module 'tsconfck' or its corresponding type declarations.

4 import {parse as tsparse} from 'tsconfck'
                                 ~~~~~~~~~~


Found 1 error in src/util/fs.ts:4

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/pack for documentation about this command.
```

</details>

Presumably this inability to resolve the module is also occurring when we try to do the dynamic import in the original code.

**To Reproduce**
Variant 1:
1. Apply the patch provided above to the repo.
2. Run `yarn pack`

Variant 2:
1. Run `npx ts-node` in a clone of oclif/core after having run `yarn install --frozen-lockfile`.
2. Sanity test by running `import {basename, dirname, join} from 'node:path'` to see that you can import modules.
3. Run `import {parse} from 'tsconfck'` and see it error out.

<details>
<summary>
ts-node output
</summary>

```
$ npx ts-node
> import {basename, dirname, join} from 'node:path'
undefined
> import {parse} from 'tsconfck'
<repl>.ts:5:21 - error TS2307: Cannot find module 'tsconfck' or its corresponding type declarations.

5 import {parse} from 'tsconfck'
```

</details>

**Expected behavior**
The library should be able to be found.

**Environment (please complete the following information):**
 - OS & version: MacOS Venture 13.6.3
 - Shell/terminal & version:  bash-5.2.21
 - Node version: 18.18.2
 - Yarn version: 1.22.21

**Additional context**
Looks like folks are having other problems with tsconfck as well: https://github.com/oclif/core/issues/878.

Looks like tsconfck removed support for cjs exports: https://github.com/dominikg/tsconfck/releases/tag/tsconfck%403.0.0

### PR DESCRIPTION
**Pull request by mdonnalley**
Drop use of `tsconfck` library for resolving extended tsconfigs. 

Fixes #878 
Fixes #907 
@W-14526758@

### CODE REVIEW

File: src/flags.ts [hunk 1]
```diff
@@ -99,6 +115,11 @@ export const directory = custom<string, {exists?: boolean}>({
   },
 })
 
+/**
+ * A flag flag.
```
Comment (iowillhoit) [inline src/flags.ts:119]:
```suggestion
 * A file flag.
```

**Comment by iowillhoit** [comment]
QA Looks good. Nice work on the integration tests, I cannot think of another case that we would need to cover. 

Sidenote: Do we want to account for the fact that `references` are excluded from inheritance ([docs](https://www.typescriptlang.org/tsconfig#references:~:text=Currently%2C%20the%20only%20top%2Dlevel%20property%20that%20is%20excluded%20from%20inheritance%20is%20references.))? It looks like this is how it has been working for a while now (before this PR)

**Review by iowillhoit — APPROVED** [review]

File: src/util/fs.ts [hunk 2]
```diff
@@ -71,10 +69,3 @@ export async function safeReadJson<T>(path: string): Promise<T | undefined> {
 export function existsSync(path: string): boolean {
   return fsExistsSync(path)
 }
-
-export async function readTSConfig(path: string) {
-  const {parse} = await import('tsconfck')
-  const result = await parse(path)
-  const tsNodeOpts = mergeNestedObjects(result.extended ?? [result], 'tsconfig.ts-node')
-  return {...result.tsconfig, 'ts-node': tsNodeOpts}
-}
```
[the remark this answers is not in the document]
Comment (mdonnalley) [reply src/util/fs.ts:80]: It was never export from [src/index.ts](https://github.com/oclif/core/blob/main/src/index.ts) so we're able to remove it without it being a breaking change

**Comment by mdonnalley** [comment]
>Sidenote: Do we want to account for the fact that references are excluded from inheritance ([docs](https://www.typescriptlang.org/tsconfig#references:~:text=Currently%2C%20the%20only%20top%2Dlevel%20property%20that%20is%20excluded%20from%20inheritance%20is%20references.))? It looks like this is how it has been working for a while now (before this PR)

I would wait until someone needs this before taking the time. I generally like to avoid adding support for something that there's not an immediate use case for

File: src/util/fs.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (iowillhoit) [reply src/util/fs.ts:80]: Good point. They could grab the export from this file but what we export from src/index is our "contract".

**Pull request merged**